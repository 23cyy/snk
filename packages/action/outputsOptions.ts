import { Options as DrawOptions } from "@snk/svg-creator";
import { palettes } from "./palettes";

export const parseOutputsOption = (lines: string[]) => lines.map(parseEntry);

export const parseEntry = (entry: string) => {
  const m = entry.trim().match(/^(.+\.(svg|gif))(\?.*)?$/);
  if (!m) return null;

  const [_, filename, format, query] = m;

  const sp = new URLSearchParams(query || "");

  const drawOptions: DrawOptions = {
    sizeDotBorderRadius: 2,
    sizeCell: 16,
    sizeDot: 12,
    ...palettes["default"],
  };
  const gifOptions = { step: 1, frameDuration: 100 };

  {
    const palette = palettes[sp.get("palette")!];
    if (palette) {
      Object.assign(drawOptions, palette);
      drawOptions.dark = palette.dark && { ...palette.dark };
    }
  }

  if (sp.has("color_snake")) drawOptions.colorSnake = sp.get("color_snake")!;
  if (sp.has("color_dots")) {
    const colors = sp.get("color_dots")!.split(/[,;]/);
    drawOptions.colorDots = colors;
    drawOptions.colorEmpty = colors[0];
    drawOptions.dark = undefined;
  }
  if (sp.has("color_dot_border"))
    drawOptions.colorDotBorder = sp.get("color_dot_border")!;

  if (sp.has("dark_color_dots")) {
    const colors = sp.get("dark_color_dots")!.split(/[,;]/);
    drawOptions.dark = {
      ...drawOptions.dark,
      colorDots: colors,
      colorEmpty: colors[0],
    };
  }
  if (sp.has("dark_color_dot_border") && drawOptions.dark)
    drawOptions.dark.colorDotBorder = sp.get("color_dot_border")!;
  if (sp.has("dark_color_snake") && drawOptions.dark)
    drawOptions.dark.colorSnake = sp.get("color_snake")!;

  return { filename, format: format as "svg" | "gif", drawOptions, gifOptions };
};
