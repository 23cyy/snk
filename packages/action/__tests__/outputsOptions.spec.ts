import { parseEntry } from "../outputsOptions";

[
  "path/to/out.gif",

  "/out.svg?color_snake=orange&color_dots=#bfd6f6,#8dbdff,#64a1f4,#4b91f1,#3c7dd9",
].forEach((entry) =>
  it(`should parse ${entry}`, () => {
    expect(parseEntry(entry)).toMatchSnapshot();
  })
);
