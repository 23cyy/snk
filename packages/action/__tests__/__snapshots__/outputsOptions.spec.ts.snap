// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse /out.svg?color_snake=orange&color_dots=#bfd6f6,#8dbdff,#64a1f4,#4b91f1,#3c7dd9 1`] = `
Object {
  "drawOptions": Object {
    "colorDotBorder": "#1b1f230a",
    "colorDots": Array [
      "#bfd6f6",
      "#8dbdff",
      "#64a1f4",
      "#4b91f1",
      "#3c7dd9",
    ],
    "colorEmpty": "#bfd6f6",
    "colorSnake": "orange",
    "dark": undefined,
    "sizeCell": 16,
    "sizeDot": 12,
    "sizeDotBorderRadius": 2,
  },
  "filename": "/out.svg",
  "format": "svg",
  "gifOptions": Object {
    "frameDuration": 100,
    "step": 1,
  },
}
`;

exports[`should parse path/to/out.gif 1`] = `
Object {
  "drawOptions": Object {
    "colorDotBorder": "#1b1f230a",
    "colorDots": Array [
      "#ebedf0",
      "#9be9a8",
      "#40c463",
      "#30a14e",
      "#216e39",
    ],
    "colorEmpty": "#ebedf0",
    "colorSnake": "purple",
    "dark": Object {
      "colorDotBorder": "#1b1f230a",
      "colorDots": Array [
        "#161b22",
        "#01311f",
        "#034525",
        "#0f6d31",
        "#00c647",
      ],
      "colorEmpty": "#161b22",
      "colorSnake": "purple",
    },
    "sizeCell": 16,
    "sizeDot": 12,
    "sizeDotBorderRadius": 2,
  },
  "filename": "path/to/out.gif",
  "format": "gif",
  "gifOptions": Object {
    "frameDuration": 100,
    "step": 1,
  },
}
`;
